src/java/org/apache/cassandra/gms/EndpointState.java

```diff
@@ -164,7 +185,13 @@ public synchronized void addApplicationStates(Set<Map.Entry<ApplicationState, Ve
 
             if (applicationState.compareAndSet(orig, copy))
             {
-                update(copy.entrySet());
+                EnumMap<ApplicationState, VersionedValue> diff = new EnumMap<ApplicationState, VersionedValue>(copy);
+                for (Map.Entry<ApplicationState, VersionedValue> entry : copy.entrySet())
+                {
```

Comment:
this looks weird to me...

we just updated `copy` using `values` and now we do kind of the same thing to figure out the diff....
perhaps it makes more sense to do it in one loop? And comparing versions, not whole objects? WDYT?